packages/realm/src/app-services/SubscriptionSet.ts

```diff
@@ -0,0 +1,120 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2023 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+import {
+  BaseSubscriptionSet,
+  FlexibleSyncConfiguration,
+  MutableSubscriptionSet,
+  Realm,
+  SubscriptionsState,
+  assert,
+  binding,
+} from "../internal";
+
+/**
+ * Class representing the set of all active flexible sync subscriptions for a Realm
+ * instance.
+ *
+ * The server will continuously evaluate the queries that the instance is subscribed to
+ * and will send data that matches them, as well as remove data that no longer does.
+ *
+ * The set of subscriptions can only be updated inside a {@link SubscriptionSet.update} callback,
+ * by calling methods on the corresponding {@link MutableSubscriptionSet} instance.
+ */
+export class SubscriptionSet extends BaseSubscriptionSet {
+  /**@internal */
+  constructor(/**@internal */ private realm: Realm, /**@internal */ internal: binding.SyncSubscriptionSet) {
+    super(internal);
+  }
+
+  /**
+   * Wait for the server to acknowledge this set of subscriptions and return the
+   * matching objects.
+   *
+   * If `state` is {@link SubscriptionsState.Complete}, the promise will be resolved immediately.
+   *
+   * If `state` is {@link SubscriptionsState.Error}, the promise will be rejected immediately.
+   *
+   * @returns A promise which is resolved when synchronization is complete, or is
+   *  rejected if there is an error during synchronisation.
+   */
+  async waitForSynchronization(): Promise<void> {
+    try {
+      const state = await this.internal.getStateChangeNotification(binding.SyncSubscriptionSetState.Complete);
+      if (state === binding.SyncSubscriptionSetState.Error) {
+        throw new Error(this.error || "Encountered an error when waiting for synchronization.");
+      }
+    } finally {
+      if (!this.realm.isClosed) {
+        this.internal.refresh();
+      }
+    }
+  }
```

Comment:
The [previous implementation](https://github.com/realm/realm-js/blob/20405a00cca1f24bdff2023bbd914bbc8a7e4f33/src/js_subscriptions.hpp#L522-L527) caught a `KeyNotFound` exception, but I don't think this is needed anymore.

Reply:
https://github.com/realm/realm-core/issues/5165 has been closed so I assume we are fine.